### PR TITLE
fix cursor position in case of suggested suffix

### PR DESCRIPTION
**Pull request by dimacodota**
- Substitute command was removed - the purpose of this command was to handle the cursor in case of new suffix suggested by tabnine, replacement of the old prefix and suffix, and position the cursor before the suggested new suffix for all selections. instead, [SnippetString](https://code.visualstudio.com/api/references/vscode-api#SnippetString) usage was added to the competition item creation (`makeCompletionItem`), which allows to add a literal template snippet that can handle the cursor position (`appendTabstop`) between the inserted text values .
 - package-lock was updated due to deprecate usage of the flatmap-stream library (error while npm install : Not Found: flatmap-stream@https://registry.npmjs.org/flatmap-stream/-/flatmap-stream-0.1.0.tgz)

**after this fix:**

p4EjWM3ra8
LJcOS2Vmy1

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -15,54 +15,9 @@ import {Mutex} from 'await-semaphore';
 const CHAR_LIMIT = 100000;
 const MAX_NUM_RESULTS = 5;
 const DEFAULT_DETAIL = "TabNine";
-const COMMAND_NAME = "TabNine Substitute";
 
 export function activate(context: vscode.ExtensionContext) {
 
-  vscode.commands.registerTextEditorCommand(
```
Comment (dimacodota) [inline src/extension.ts]: this command was removed,  [SnippetString](https://code.visualstudio.com/api/references/vscode-api#SnippetString) was used instead

File: src/extension.ts [hunk 2]
```diff
@@ -265,13 +211,6 @@ export function activate(context: vscode.ExtensionContext) {
   }
 }
 
-interface CommandArgs {
```
Comment (dimacodota) [inline src/extension.ts]: unused interface removed

File: src/extension.ts [hunk 3]
```diff
@@ -185,22 +140,13 @@ export function activate(context: vscode.ExtensionContext) {
...
-      title: "accept completion",
-    };
-    item.insertText = args.entry.new_prefix;
-    item.range = new vscode.Range(args.position.translate(0, -args.old_prefix.length), args.position);
+    
+    item.insertText = new vscode.SnippetString(args.entry.new_prefix)
+      .appendTabstop()
+      .appendText(args.entry.new_suffix);
+
+    item.range = new vscode.Range(args.position.translate(0, -args.old_prefix.length), args.position.translate(0, args.entry.old_suffix.length));
```
Comment (dimacodota) [inline src/extension.ts]: the range was changed from the begging of the old prefix (`args.old_prefix.length`) to old suffix (`args.entry.old_suffix.length`)

File: src/extension.ts [hunk 4]
```diff
@@ -185,22 +140,13 @@ export function activate(context: vscode.ExtensionContext) {
...
-    item.command = {
-      arguments: [arg],
-      command: COMMAND_NAME,
-      title: "accept completion",
-    };
-    item.insertText = args.entry.new_prefix;
-    item.range = new vscode.Range(args.position.translate(0, -args.old_prefix.length), args.position);
+    
+    item.insertText = new vscode.SnippetString(args.entry.new_prefix)
+      .appendTabstop()
```
Comment (dimacodota) [inline src/extension.ts]: appendTabstop() positions the cursor after the new prefix and before new suffix

File: src/extension.ts [hunk 5]
```diff
@@ -185,22 +140,13 @@ export function activate(context: vscode.ExtensionContext) {
     : vscode.CompletionItem
   {
     let item = new vscode.CompletionItem(args.entry.new_prefix);
-    item.insertText = "";
```
Comment (dimacodota) [inline src/extension.ts]: duplicate initialization was removed

**Comment by dimacodota** [comment]
> LGTM. Seems to be quite a hefty change, we should test for regression

agree

File: src/extension.ts (hunk 3 above)
Comment (zxqfl) [inline src/extension.ts]: One thing to check here is that the text after the cursor is what we expect it to be. The most likely way for it to not be what we expect is if the user has multiple selections, and the text after the cursor is different in the multiple selections. Here's an example of a case where we should make sure not to replace the `..` after the second `foo`:
image

Comment (dimacodota) [reply src/extension.ts]: hey Jacob, thanks for bringing this to my attention, this is actually the range property behaviour, it operates only on the main selection

multiselct-cursor position

**Comment by dimacodota** [comment]
@avichay77 FYI

**Pull request merged**